### PR TITLE
Andy/sqldiffs

### PR DESCRIPTION
**Pull request by andrew-wm-arthur**
Created option to output diffs as SQL queries. `dolt diff --sql` & `dolt diff -q`
First version only supported data diffs. If schema diffs exist, the command will exit

### CODE REVIEW

File: go/cmd/dolt/commands/diff.go [hunk 1]
```diff
@@ -84,16 +94,18 @@ var diffSynopsis = []string{
...
+	limit      int
+	where      string
 }
 
 func Diff(ctx context.Context, commandStr string, args []string, dEnv *env.DoltEnv) int {
 	ap := argparser.NewArgParser()
 	ap.SupportsFlag(DataFlag, "d", "Show only the data changes, do not show the schema changes (Both shown by default).")
 	ap.SupportsFlag(SchemaFlag, "s", "Show only the schema changes, do not show the data changes (Both shown by default).")
 	ap.SupportsFlag(SummaryFlag, "", "Show summary of data changes")
+	ap.SupportsFlag(SQLFlag, "q", "Output changes as an SQL modification")
```
Comment (andrew-wm-arthur) [inline go/cmd/dolt/commands/diff.go]: This isn't the right way to say this. Any suggestions?

File: go/libraries/doltcore/table/untyped/sqlexport/sqlwriter.go [hunk 2]
```diff
@@ -145,6 +160,98 @@ func (w *SqlExportWriter) insertStatementForRow(r row.Row) (string, error) {
...
+		col := w.sch.GetAllCols().TagToCol[tag]
+		if col.IsPartOfPK {
+			if seenOne {
+				b.WriteString(" AND ")
+			}
+			sqlString, err := w.sqlString(val)
+			if err != nil {
+				return true, err
+			}
+			b.WriteString(col.Name)
```
Comment (zachmu) [inline go/libraries/doltcore/table/untyped/sqlexport/sqlwriter.go]: Some column names will not be valid sql identifiers (will include spaces, etc.). You'll want to escape all column names, here and elsewhere with backticks (`colName`).

Also add a test to ensure this works

Comment (zachmu) [reply go/libraries/doltcore/table/untyped/sqlexport/sqlwriter.go]: Looks like I already wrote sql.QuoteIdentifier for this purpose

File: go/libraries/doltcore/table/untyped/sqlexport/sqlwriter.go [hunk 3]
```diff
@@ -145,6 +160,98 @@ func (w *SqlExportWriter) insertStatementForRow(r row.Row) (string, error) {
...
+		col := w.sch.GetAllCols().TagToCol[tag]
+		if !col.IsPartOfPK {
+			if seenOne {
+				b.WriteRune(',')
+			}
+			sqlString, err := w.sqlString(val)
+			if err != nil {
+				return true, err
+			}
+			b.WriteString(col.Name)
```
Comment (zachmu) [inline go/libraries/doltcore/table/untyped/sqlexport/sqlwriter.go]: Same comment about escaping

File: go/libraries/doltcore/table/untyped/sqlexport/sqlwriter.go [hunk 4]
```diff
@@ -145,6 +160,98 @@ func (w *SqlExportWriter) insertStatementForRow(r row.Row) (string, error) {
...
+		col := w.sch.GetAllCols().TagToCol[tag]
+		if col.IsPartOfPK {
+			if seenOne {
+				b.WriteString(" AND ")
+			}
+			sqlString, err := w.sqlString(val)
+			if err != nil {
+				return true, err
+			}
+			b.WriteString(col.Name)
```
Comment (zachmu) [inline go/libraries/doltcore/table/untyped/sqlexport/sqlwriter.go]: Escaping

File: go/libraries/doltcore/table/untyped/sqlexport/sqlwriter.go [hunk 5]
```diff
@@ -145,6 +160,98 @@ func (w *SqlExportWriter) insertStatementForRow(r row.Row) (string, error) {
 	return b.String(), nil
 }
 
+func (w *SqlExportWriter) WriteDeleteRow(ctx context.Context, r row.Row) error {
```
Comment (zachmu) [inline go/libraries/doltcore/table/untyped/sqlexport/sqlwriter.go]: You should update sqlwriter_test.go with simple tests for the new methods

File: bats/sql_diff.bats [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+    [[ "$output" != "" ]] || false
+
+    dolt diff --sql newbranch firstbranch > query
+    dolt checkout firstbranch
+    dolt sql < query
+    rm query
+    dolt add test
+    dolt commit -m "Reconciled with newbranch"
+
+    # confirm difference was reconciled
```
[the remark this answers is not in the document]
Comment (andrew-wm-arthur) [reply bats/sql_diff.bats]: done

File: go/cmd/dolt/commands/diff.go [hunk 7]
```diff
@@ -54,13 +54,23 @@ const (
 
 	SchemaAndDataDiff = SchemaOnlyDiff | DataOnlyDiff
 
+	ColorDiffOutput = 1
```
[the remark this answers is not in the document]
Comment (andrew-wm-arthur) [reply go/cmd/dolt/commands/diff.go]: done

File: go/cmd/dolt/commands/diff.go [hunk 8]
```diff
@@ -84,16 +94,18 @@ var diffSynopsis = []string{
 }
 
 type diffArgs struct {
-	diffParts int
-	limit     int
-	where     string
+	diffParts  int
+	diffOutput int
```
[the remark this answers is not in the document]
Comment (andrew-wm-arthur) [reply go/cmd/dolt/commands/diff.go]: done

File: go/cmd/dolt/commands/diff.go [hunk 9]
```diff
@@ -344,12 +363,16 @@ func diffRoots(ctx context.Context, r1, r2 *doltdb.RootValue, tblNames []string,
 			verr = diffSummary(ctx, rowData1, rowData2, colLen)
 		}
 
+		if sch1Hash != sch2Hash && dArgs.diffOutput&SQLDiffOutput != 0 {
```
[the remark this answers is not in the document]
Comment (andrew-wm-arthur) [reply go/cmd/dolt/commands/diff.go]: done

File: go/libraries/doltcore/diff/sql_diff_sink.go [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+		if dt, convertedOK := prop.(DiffChType); convertedOK {
+			switch dt {
+			case DiffAdded:
+				return sds.sw.WriteInsertRow(context.TODO(), r)
+			case DiffRemoved:
+				return sds.sw.WriteDeleteRow(context.TODO(), r)
+			case DiffModifiedOld:
+				return nil
+			case DiffModifiedNew:
+				return sds.sw.WriteUpdateRow(context.TODO(), r)
```
[the remark this answers is not in the document]
Comment (andrew-wm-arthur) [reply go/libraries/doltcore/diff/sql_diff_sink.go]: done

File: go/libraries/doltcore/table/untyped/sqlexport/sqlwriter.go (hunk 2 above)
Comment (andrew-wm-arthur) [reply go/libraries/doltcore/table/untyped/sqlexport/sqlwriter.go]: done

File: go/libraries/doltcore/table/untyped/sqlexport/sqlwriter.go (hunk 5 above)
Comment (andrew-wm-arthur) [reply go/libraries/doltcore/table/untyped/sqlexport/sqlwriter.go]: done

File: go/libraries/doltcore/table/untyped/sqlexport/sqlwriter.go (hunk 3 above)
Comment (andrew-wm-arthur) [reply go/libraries/doltcore/table/untyped/sqlexport/sqlwriter.go]: done

File: go/libraries/doltcore/table/untyped/sqlexport/sqlwriter.go (hunk 4 above)
Comment (andrew-wm-arthur) [reply go/libraries/doltcore/table/untyped/sqlexport/sqlwriter.go]: done

File: go/libraries/doltcore/doltdb/root_val.go [hunk 11]
```diff
@@ -299,6 +299,7 @@ func (root *RootValue) HashOf() (hash.Hash, error) {
 // TableDiff returns the slices of tables added, modified, and removed when compared with another root value.  Tables
 // In this instance that are not in the other instance are considered added, and tables in the other instance and not
 // this instance are considered removed.
+// TODO: is this logic duplicitous of doltcore/diff ?
```
[the remark this answers is not in the document]
Comment (andrew-wm-arthur) [reply go/libraries/doltcore/doltdb/root_val.go]: I think I was going for *duplicative*, but... yeah

File: go/libraries/doltcore/diff/sql_diff_sink.go [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+	sw  *sqlexport.SqlExportWriter
+}
+
+func NewSQLDiffSink(wr io.WriteCloser, sch schema.Schema, tableName string) (*SQLDiffSink, error) {
+	sw, err := sqlexport.NewSQLDiffWriter(wr, tableName, sch)
+
+	if err != nil {
+		return nil, err
+	}
+	//sw.SetWrittenFirstRow(true)
```
[the remark this answers is not in the document]
Comment (andrew-wm-arthur) [reply go/libraries/doltcore/diff/sql_diff_sink.go]: done

File: go/cmd/dolt/commands/diff.go [hunk 13]
```diff
@@ -575,14 +605,17 @@ func buildPipeline(dArgs *diffArgs, joiner *rowconv.Joiner, ds *diff.DiffSplitte
...
 	nullPrinter := nullprinter.NewNullPrinter(untypedUnionSch)
-
-	transforms.AppendTransforms(pipeline.NewNamedTransform("split_diffs", ds.SplitDiffIntoOldAndNew),
+	transforms.AppendTransforms(
+		pipeline.NewNamedTransform("split_diffs", ds.SplitDiffIntoOldAndNew),
 		pipeline.NewNamedTransform(nullprinter.NULL_PRINTING_STAGE, nullPrinter.ProcessRow),
-		pipeline.NamedTransform{Name: fwtStageName, Func: fwtTr.TransformToFWT},
 	)
 
+	if dArgs.diffOutput&ColorDiffOutput != 0 {
```
[the remark this answers is not in the document]
Comment (andrew-wm-arthur) [reply go/cmd/dolt/commands/diff.go]: done

File: go/cmd/dolt/commands/diff.go [hunk 14]
```diff
@@ -516,29 +544,31 @@ func diffRows(ctx context.Context, newRows, oldRows types.Map, newSch, oldSch sc
 		return verr
 	}
 
-	if schemasEqual {
-		schRow, err := untyped.NewRowFromTaggedStrings(newRows.Format(), untypedUnionSch, newColNames)
+	if dArgs.diffOutput&SQLDiffOutput == 0 {
```
[the remark this answers is not in the document]
Comment (andrew-wm-arthur) [reply go/cmd/dolt/commands/diff.go]: done

File: go/cmd/dolt/commands/diff.go [hunk 15]
```diff
@@ -497,7 +520,12 @@ func diffRows(ctx context.Context, newRows, oldRows types.Map, newSch, oldSch sc
 		numHeaderRows = 2
 	}
 
-	sink, err := diff.NewColorDiffSink(iohelp.NopWrCloser(cli.CliOut), untypedUnionSch, numHeaderRows)
+	var sink DiffSink
+	if dArgs.diffOutput&ColorDiffOutput != 0 {
```
[the remark this answers is not in the document]
Comment (andrew-wm-arthur) [reply go/cmd/dolt/commands/diff.go]: done

File: go/cmd/dolt/commands/diff.go [hunk 16]
```diff
@@ -84,16 +97,18 @@ var diffSynopsis = []string{
...
+	limit      int
+	where      string
 }
 
 func Diff(ctx context.Context, commandStr string, args []string, dEnv *env.DoltEnv) int {
 	ap := argparser.NewArgParser()
 	ap.SupportsFlag(DataFlag, "d", "Show only the data changes, do not show the schema changes (Both shown by default).")
 	ap.SupportsFlag(SchemaFlag, "s", "Show only the schema changes, do not show the data changes (Both shown by default).")
 	ap.SupportsFlag(SummaryFlag, "", "Show summary of data changes")
+	ap.SupportsFlag(SQLFlag, "q", "Output diff as a SQL patch file of INSERT/ UPDATE / DELETE statements")
```
Comment (zachmu) [inline go/cmd/dolt/commands/diff.go]: Missing a space after INSERT

File: go/libraries/doltcore/table/untyped/sqlexport/sqlwriter_test.go [hunk 17]
```diff
@@ -121,6 +121,102 @@ func TestWriteRow(t *testing.T) {
...
+		schema.NewColumn("anotherCol", 0, types.FloatKind, false),
+		schema.NewColumn("a name with spaces", 1, types.IntKind, true),
+	)
+
+	tests := []test{
+		{
+			name:           "negative values and columns with spaces",
+			rows:           rs(dtestutils.NewRow(trickySch, types.Float(-3.14), types.Int(-42))),
+			sch:            trickySch,
+			expectedOutput: "DELETE FROM `tricky`  WHERE (`a name with spaces`=-42);" + "\n",
```
Comment (zachmu) [inline go/libraries/doltcore/table/untyped/sqlexport/sqlwriter_test.go]: Looks like you have an extra space before WHERE here

**Pull request merged**